functions/src/index.ts

```diff
@@ -57,6 +75,118 @@ export const finishGame = functions.https.onCall(async (data, context) => {
+    for (const player of players) {
+      scores[player] ??= 0;
+    }
+
+    // Retrieve old ratings from the database.
+    const ratings: Record<string, number> = {};
+    for (const player of players) {
+      const ratingSnap = await admin
+        .database()
+        .ref(`userStats/${player}/${gameMode}/rating`)
+        .once("value");
+      const rating = ratingSnap.exists() ? ratingSnap.val() : BASE_RATING;
+      ratings[player] = rating;
+    }
+
+    // Translate ratings to exponential format.
+    const expRatings: Record<string, number> = {};
+    for (const player of players) {
+      expRatings[player] = Math.pow(10, ratings[player] / SCALING_FACTOR);
+    }
+
+    // Compute expected ratio for each player.
+    const expectedRatio: Record<string, number> = {};
+    let ratingSum = 0;
+    for (const player of players) {
+      ratingSum += expRatings[player];
+    }
+    for (const player of players) {
+      expectedRatio[player] = expRatings[player] / ratingSum;
+    }
+
+    // Compute achieved ratio for each player.
+    const achievedRatio: Record<string, number> = {};
+    for (const player of players) {
+      achievedRatio[player] = scores[player] / setCount;
+    }
+
+    // Compute new rating for each player.
+    const playerCountMultiplier = 1.0 / (players.length - 1);
+    // const gameCounts: Record<string, number> = {};
+    for (const player of players) {
+      let learningRate = LEARNING_RATE_BEGINNER;
+
+      /**
+       * This code is currently not in use. It has been written to accommodate
+       * adapting the learning rate to the experience of a player and the
+       * corresponding certainty of the rating system based on the number of
+       * games played. It is specifically aimed for use after a restructuring of
+       * the database where the users statistics are moved over to the DB.
+       */
+      // const finishedGamesCountSnap = await admin
+      //   .database()
+      //   .ref(`userStats/${player}/${gameMode}/finishedGamesCount`)
+      //   .once("value");
+      // gameCounts[player] = finishedGamesCountSnap.exists()
+      //   ? finishedGamesCountSnap.val()
+      //   : 0;
+      //
+      // if (gameCounts[player] >= LEARNING_RATE_ADVANCED_THRESHOLD) {
+      //   learningRate = LEARNING_RATE_ADVANCED;
+      // } else if (gameCounts[player] >= LEARNING_RATE_INTERMEDIATE_THRESHOLD) {
+      //   learningRate = LEARNING_RATE_INTERMEDIATE;
+      // }
+
+      const newRating =
+        ratings[player] +
+        playerCountMultiplier *
+          learningRate *
+          (achievedRatio[player] - expectedRatio[player]);
+      ratings[player] = newRating;
+    }
+
+    // Push new ratings to the database.
+    const updates: Array<Promise<any>> = [];
+    for (const player of players) {
+      updates.push(
+        admin.database().ref(`userStats/${player}/${gameMode}`).update({
+          rating: ratings[player],
+        })
+      );
+    }
+
+    await Promise.all(updates);
```

Comment:
I think you're accidentally updating things too many times. This is in another nested loop that updates every player's rating n times in an n-player game.